uaclient/security.py

```diff
@@ -465,69 +498,109 @@ def prompt_for_affected_packages(
+        token = input("> ")
+        print(status.colorize_commands([["ua", "attach", token]]))
+        return bool(
+            0
+            == cli.action_attach(
+                argparse.Namespace(token=token, auto_enable=True), cfg
+            )
+        )
+    elif choice == "s":
+        print(status.PROMPT_UA_SUBSCRIPTION_URL)
+        print("TODO: GH: #1413: magic subscription attach")
+    return True
 
 
 def upgrade_packages_and_attach(
     cfg: UAConfig,
     upgrade_packages: "List[str]",
     upgrade_packages_ua: "List[str]",
-):
+) -> bool:
     """Upgrade available packages to fix a CVE.
 
     Upgrade all packages in Ubuntu standard updates regardless of attach
     status.
 
     For UA upgrades, prompt regarding system attach prior to upgrading UA
     packages.
-    """
-    if upgrade_packages:
-        if os.getuid() != 0:
-            print(status.MESSAGE_SECURITY_APT_NON_ROOT)
-            return
-
-        apt.run_apt_command(
-            cmd=["apt-get", "update"],
-            error_msg=status.MESSAGE_APT_UPDATE_FAILED,
-            print_cmd=True,
-        )
 
-        apt.run_apt_command(
-            cmd=["apt-get", "install", "--only-upgrade", "-y"]
-            + upgrade_packages,
-            error_msg=status.MESSAGE_APT_INSTALL_FAILED,
-            env={"DEBIAN_FRONTEND": "noninteractive"},
-            print_cmd=True,
-        )
+    :return: True if package upgrade completed or unneeded, False otherwise.
+    """
+    if not any([upgrade_packages, upgrade_packages_ua]):
+        return True
+    if os.getuid() != 0:
+        print(status.MESSAGE_SECURITY_APT_NON_ROOT)
+        return False
     if upgrade_packages_ua:
-        if cfg.is_attached:
-            print(
-                "TODO: GH: #1402: apt commands to install missing UA updates"
-            )
-        else:
-            print(status.MESSAGE_SECURITY_UPDATE_NOT_INSTALLED_SUBSCRIPTION)
-            choice = util.prompt_choices(
-                "Choose: [S]ubscribe at ubuntu.com [A]ttach"
-                " existing token [C]ancel",
-                valid_choices=["s", "a", "c"],
-            )
-            print("TODO react to subscription choice:{} GH: #".format(choice))
+        if not cfg.is_attached:
+            if not _prompt_for_attach(cfg):
+                return False  # User opted to cancel
+    packages = sorted(upgrade_packages + upgrade_packages_ua)
+    print(
+        status.colorize_commands(
+            [
+                ["apt", "update", "&&"]
+                + ["apt", "install", "--only-upgrade", "-y"]
+                + packages
+            ]
+        )
+    )
+    apt.run_apt_command(
+        cmd=["apt-get", "update"], error_msg=status.MESSAGE_APT_UPDATE_FAILED
+    )
+    apt.run_apt_command(
+        cmd=["apt-get", "install", "--only-upgrade", "-y"] + packages,
+        error_msg=status.MESSAGE_APT_INSTALL_FAILED,
+        env={"DEBIAN_FRONTEND": "noninteractive"},
+        print_cmd=True,
```

Comment:
Since we are already printing the commands, we should drop the `print_cmd`  part from this function

Reply:
I would vote +1 to drop per my above response but I wanted to check with you today for your concerns here https://github.com/canonical/ubuntu-advantage-client/pull/1414#discussion_r586620039